[WIP] Forking glog into third_party/ directory

dims opened the pull request:
Change-Id: I9756b73c806f8e29edabae3999c06d4083a6527b

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

- Move the glog directory to third_party/forked
- patch up godep-save.sh to make sure links are created
- patch up hack/update-bazel.sh so BUILD is correct
- patch up update-godep-licenses.sh so LICENSES get generated correctly
- patch hack/verify-pkg-names.sho to skip the forked glog directory

Forking glog is a first step to essentially get rid of it, please see discussion in #61006 and elsewhere on why glog id problematic

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

dims committed:
Forking glog go files into third_party/ directory

- fork code into third_party/forked/glog/ directory
- patch up hack/godep-save.sh to make sure links are created
- patch hack/verify-pkg-names.sho to skip the forked glog directory
- link vendor/github.com/golang/glog/glog.go and glog_file.go from the
  forked directory

Change-Id: Ife852c6e8f868ba157a4d6c3dc7c66c1bfac91e7

dims commented:
/test pull-kubernetes-bazel-test

dims commented:
related to https://github.com/kubernetes/kubernetes/issues/61006

neolit123 commented:
would be quite nice if this works! :+1:

tallclair commented:
@dims I'd like to be able to take advantage of the changes we make in non-k/k repos (in particular, to fix all the container images that require a shell just to redirect the logs). Will it be possible to vendor the fork from here, or should we consider putting it elsewhere (perhaps it's own repo?)

dims commented:
@tallclair my thought was to iterate here in k/k first, then put the fork in its own repo and get the projects in our dependencies to switch over. 

for example, we need to take care of problems with say `"log_dir"` causing issues when there are multiple copies of the code.

dims commented:
/kind cleanup

dims commented:
/test pull-kubernetes-integration

dims commented:
/test pull-kubernetes-node-e2e

dims commented:
@tallclair i've switched things around to be able to vendor the fork from here, also in the process of adding a `log-file` parameter as well

justinsb commented:
Well this is an interesting idea :-)

I do think we should improve our logging, but I think we should decide how far we're going to go:

* we could just fix the problems with glog, but essentially keep the same approach
* we could adopt more structured logging
* we could also add tracing (OpenTracing / OpenCensus).  The reason I'm conflating them is because many log messages are summarizing a tracing scope enter/exit, and often contextual information that would be logged is in the tracing scope.  I think considering them separately is going to logging and tracing that are very duplicitative of each other. 

My 2c: we should take this to sig-architecture (have we done so already?), and prototype this somewhere outside of k/k.  I'd normally suggest kops, but kops is mostly not controllers & servers, so is not a very representative codebase.

tallclair commented:
@justinsb If you haven't already, see https://github.com/kubernetes/kubernetes/issues/61006
I think the idea here was to fix a few of the immediate problems with glog, and give us a smooth migration path once we have settled on a better log library.

dims committed:
Add option log-file

- Move all flag initialization to a single init() block
- add a new option `log-file`
- Avoid creating directories when `log-file` is specified

Change-Id: I04ef706af9260902aacf43a38ed75ba06df1449d

dims committed:
Ability to redirect glog output somewhere else

Change-Id: Id5a97c61d5f611959a284e1fa0ca79091495dd02

dims committed:
Example redirecting log/glog to stderr

Change-Id: I010ebb01483bfb53b532646047d959ce6c7d0579

dims committed:
Drop the logtostderr hack

Change-Id: I95f1bbdeaf9c993c704a7984a4ad4770436cc8fe

dims commented:
/assign @thockin

dims committed:
Now try writing all the logs to syslog

Change-Id: Iecf92407333123d6427476edcd093b9df99fc3e8

dims committed:
Add go-logr/glogr and go-logr/logr to show it can co-exist

Change-Id: I80e40b20298c73f3eefe8b743ad46871af24a73b

dims commented:
/test pull-kubernetes-integration

dims commented:
/test pull-kubernetes-e2e-kubeadm-gce

thockin commented on third_party/forked/glog/glog_file.go:
```diff
@@ -36,15 +35,17 @@ var MaxSize uint64 = 1024 * 1024 * 1800
-	if *logDir != "" {
-		logDirs = append(logDirs, *logDir)
+	// if log-file is not specified, only then try to use logDir
+	if logging.logFile == "" {
```
Writing this as a switch would maybe make it easier to read

thockin commented on third_party/forked/glog/glog_file.go:
```diff
@@ -36,15 +35,17 @@ var MaxSize uint64 = 1024 * 1024 * 1800
-		logDirs = append(logDirs, *logDir)
+	// if log-file is not specified, only then try to use logDir
+	if logging.logFile == "" {
+		// if log-dir is not specified, then use temp directory
```
This is a pretty deep change - it changes "log to all dirs" into "log to one dir", if I read correctly.  Why bother to keep logDirs as a plural at this point?

thockin commented on third_party/forked/glog/glog.go:
```diff
@@ -677,6 +677,42 @@ func (l *loggingT) printWithFileLine(s severity, file string, line int, alsoToSt
 	l.output(s, buf, file, line, alsoToStderr)
 }
 
+type redirectBuffer struct {
```
comments

thockin commented on pkg/kubectl/util/logs/logs.go:
```diff
@@ -27,19 +28,20 @@ import (
 
-// GlogWriter serves as a bridge between the standard log package and the glog package.
-type GlogWriter struct{}
+// ErrorLogWriter serves as a bridge between the standard log package and the glog package.
```
Would it be better or worse to set --log-file=/dev/stderr instead of this?  It seems less invasive and maybe we could drop the SetOutput stuff ?

thockin replied:
In doing it this way, do you lose teh formatted prefix (timestamp, etc) on each line?  Or if not, and you send to another logger, do you log two timestamps?

thockin commented on pkg/kubectl/util/logs/logs.go:
```diff
@@ -28,17 +30,18 @@ import (
 
-// ErrorLogWriter serves as a bridge between the standard log package and the glog package.
-type ErrorLogWriter struct{}
+// SysLogWriter serves as a bridge between the standard log package and the glog package.
```
Is this a change we want to keep or just a proof?

thockin commented:
So, IIUC, the discussion comes down to which path to choose:

1) Do nothing, leave it all alone.

2) Replace glog's output with $other_output (stderr, logrus, zap, etc); use glog as the API for now; possibly convert to a better API $later
  pro: works now
  pro: delivers better perf/behavior ASAP with minimal churn
  con: does not address log API / structured vs printf / spam
  con: does not at address proliferation of deps on glog
  meh: ugly deps/build hacks, but not unprecedented
  meh: need to fork glog, but it's frozen anyway so low risk
 
3) Replace glog with a better API; convert to a better implementation $later
  pro: forces us to re-evaluate logging output
  pro: can set a trend for other go projects
  con: longer timeline
  con: high churn, thousands of call-sites to touch
  meh: better logging API is (maybe) less likely to be abused, but probably not

@DirectXMan12 @tallclair what else am I missing?

dims commented:
@thockin i'd like to open the door for #3 for long term and #2 for the short-term. We cannot do #3 without doing #2. We can talk more on the call

thockin commented:
I don't see how 3 depends on 2.  The two steps seem like they can be in
either order, and the major difference is which order to run with.

On Thu, Oct 25, 2018 at 10:34 AM Davanum Srinivas <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> i'd like to open the door for #3
> <https://github.com/kubernetes/kubernetes/issues/3> for long term and #2
> <https://github.com/kubernetes/kubernetes/issues/2> for the short-term.
> We cannot do #3 <https://github.com/kubernetes/kubernetes/issues/3>
> without doing #2 <https://github.com/kubernetes/kubernetes/issues/2>. We
> can talk more on the call
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/69333#issuecomment-433139728>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGzM5ZwyB_Lcsl4akNNNjIJA7fncks5uofY-gaJpZM4XEevI>
> .
>

dims commented:
we now have `k8s.io/klog` so let's use that and drop the stuff here
